### ISSUE TITLE
CGI redirection and response status code

### ISSUE THREAD

**Issue by pparadzik** [warmcat/libwebsockets#899]
Hi,

It seems that CGI redirection is not handled properly. When CGI script returns ``Location`` header, server should return ``302`` status code instead of ``200``. According to the [RFC 3875](https://tools.ietf.org/html/rfc3875#section-6.2.3): 

> The CGI script can return an absolute URI path in a Location header
   field, to indicate to the client that it should reprocess the request
   using the URI specified

and after a few lines:
  > For an HTTP client request, the server MUST generate a 302 'Found' HTTP response message.

Currently, server is ignoring any kind of a CGI response. After process is forked, server first puts ``HTTP_STATUS_OK`` in function ``lws_http_action``, and after that, it handles CGI response in ``lws_cgi_write_split_stdout_headers`` function. So, no matter what response was, status is always ``200``. Instead of this, status code should be inserted after CGI response when we determine what the status should be.

There are also other cases where CGI script would indicate to the server to act on its output (e.g [Status header](https://tools.ietf.org/html/rfc3875#section-6.3.3)), so the ``Location`` header is not the only issue.

**Comment by lws-team** [comment]
Please try the patch just pushed on master linked above.

**Comment by pparadzik** [comment]
With this patch redirection only happens if the CGI response body is not empty. If the CGI body is empty and there is only header `Location`, client receives
```
HTTP/1.1 302 Redirect
connection: close
```
and no redirection happens. For example, following script works
```
#!/bin/sh

echo "Location: http://www.google.com"
echo
echo ""
```
and following does not work
```
#!/bin/sh

echo "Location: http://www.google.com"
echo
```

**Comment by lws-team** [comment]
Okie... how about with the additional patch on top on master now?

**Comment by pparadzik** [comment]
Redirection is now working with empty body, but for some reason, if the body is left empty, it takes much longer for server to process the request. For example, if we take this script
```
#!/bin/sh

echo "Location: http://www.google.com"
echo
```
and do
```
curl -v -L 127.0.0.1:7681/<script_route>
```
then, we first get this on the client side
```
HTTP/1.1 302 Redirect
connection: close
Location: http://www.google.com
```
After about 3 or 4 seconds, server prints this
```
lwsts[18834]: 127.0.0.1 - - [18/May/2017:17:54:37 +0200] "GET /cgi-bin/luci http/1.1" 302 77 curl/7.54.0
lwsts[18834]: lws_header_table_detach: wsi 0x74b290: ah held 4s, ah.rxpos 90, ah.rxlen 90, mode/state 1 4,wsi->more_rx_waiting 0
```
after which client issues another (redirect) request. If I send SIGINT to the server after he receives the request, client immediately issues another one without waiting.

I noticed that this is also problem with scripts that are not doing redirect and returning some body. For example, take the following script
```
#!/bin/sh

echo "Status: 200 OK"
echo
echo ""
```
After doing
```
curl -v -L 127.0.0.1:7681/<script_route>
```
curl gets
```
HTTP/1.1 200 OK
connection: close
Status: 200 OK
```
and after about 4 seconds in which client is waiting for something, server prints
```
lwsts[19947]: 127.0.0.1 - - [18/May/2017:18:22:35 +0200] "GET /cgi-bin/luci http/1.1" 200 55 curl/7.54.0
lwsts[19947]: lws_header_table_detach: wsi 0x17b7290: ah held 4s, ah.rxpos 90, ah.rxlen 90, mode/state 1 4,wsi->more_rx_waiting 0
```
and curl prints
```
* Closing connection 0
```

**Comment by lws-team** [comment]
Please try one more patch-on-top in master.

**Comment by pparadzik** [comment]
Same problem as before, only now server responds in about 2 seconds instead of 4.

**Comment by lws-team** [comment]
... one more patch...

**Comment by pparadzik** [comment]
Same thing as before. So, just to be clear what I'm doing. I have this script
```
#!/bin/sh

echo "Status: 200 OK"
echo
echo ""
```
which I'm calling with curl
```
curl -v -L 127.0.0.1:7681/<route_to_the_script>
```
After that, I immediately receive this
```
HTTP/1.1 200 OK
connection: close
Status: 200 OK
```
curl then waits 2-4 seconds after which server prints this
```
lwsts[6572]: 127.0.0.1 - - [22/May/2017:11:06:20 +0200] "GET /cgi-bin/luci http/1.1" 200 54 curl/7.54.0
```
curl prints this
```
* Closing connection 0
```
and finishes with the response.

**Comment by lws-team** [comment]
I don't get any problem I can tell with this

```
#!/bin/sh

echo "Status: 200 OK"
echo -n -e "Transfer-encoding: chunked\r\n"
echo -n -e "\r\n"
echo -n -e "8\r\n"
echo -n "hello sh"
echo -n -e "0\r\n"
```

What's the underlying problem, if you give no content-length, and don't use chunking it waits a couple of seconds?  Is that an lws problem or your script should give content-length or use chunking problem?

**Comment by pparadzik** [comment]
Your script works. It seems that there is some waiting when script is not providing `Content-length`, nor `Chunked-encoding`. So, this also works
```
echo "Status: 200 OK"
echo "Content-length: 0"
echo
```
Is it possible for server to close connection when it gets EOF and/or SIGCHILD from script in cases when script is not providing `Content-length`, nor `Chunked-encoding`?

**Comment by lws-team** [comment]
That is what one of the patches does, do the close flow immediately on receipt of POLLIN + zero length read on eg stdout child wsi.  SIGCHILD is async to draining std[out|err], so if we prioritize that we can lose data by shutting down the logical cgi with unread stdout.

Real cgis like mailman and cgit emit logical http stream length management in the form of chunking or content-length because doing otherwise is incompatible with http/1.1.  For those we know synchronously we drained all stdout (because we count body bytes against cgi-reported content-length and/or parse the cgi chunking), and we react perfectly.  If your cgi is too naive to do that, then lws lazily responds to the async cgi process close to try to ensure we drained stdout/err from it, which may add a couple of seconds.
